internal/database/main_test.go

```diff
@@ -19,5 +28,203 @@ func TestMain(m *testing.M) {
+	if err != nil {
+		return "", nil, err
+	}
+	cleanup.Add(func() { os.RemoveAll(pgHost) })
+	pgData := filepath.Join(pgHost, "data")
+	env = append(env,
+		"PGHOST="+pgHost,
+		"PGDATA="+pgData,
+		"PGDATABASE=postgres",
+	)
+
+	// CI runs as root but postgres can't :O
+	if os.Getenv("CI") != "" && syscall.Getuid() == 0 {
+		sudo, err := exec.LookPath("sudo")
+		if err != nil {
+			return "", nil, err
+		}
+
+		// need the tmp dir to be owned by postgres user. Easier to use
+		// exec than looking up uid.
+		if err := exec.Command("chown", "postgres", pgHost).Run(); err != nil {
+			return "", nil, err
+		}
+
+		// need to wrap commands so they run as postgres user
+		old := cmdPreRun
+		cmdPreRun = func(c *exec.Cmd) {
+			c.Args = append([]string{"sudo", "-u", "postgres", "-E", c.Path}, c.Args[1:]...)
+			c.Path = sudo
+			old(c)
+		}
+	}
+
+	// Create the database without auth and without fsync.
+	cmd := exec.Command("initdb", pgData, "--nosync", "-E", "UNICODE", "--auth=trust")
+	cmd.Env = env
+	cmdPreRun(cmd)
+	if err := cmd.Run(); err != nil {
+		return "", nil, err
+	}
+
+	// Tune configuration for speed with ephemeral test data.
+	if err := appendFile(filepath.Join(pgData, "postgresql.conf"), []byte(`
+unix_socket_directories = '`+pgHost+`'
+listen_addresses = ''
+shared_buffers = 12MB
+fsync = off
+synchronous_commit = off
+full_page_writes = off
+log_min_duration_statement = 0
+`), 0644); err != nil {
+		return "", nil, err
+	}
+
+	// start postgres as a subprocess. If the test process dies hopefully the
+	// OS will come in and cleanup. We parse stderr looking for the log
+	// message saying it is ready.
+	cmd = exec.Command("postgres", "-D", pgData)
+	cmd.Env = env
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return "", nil, err
+	}
+	cleanup.Add(func() { stderr.Close() })
+
+	// Postgres logs with test output \o/
+	var r io.Reader = stderr
+	if testing.Verbose() {
+		r = io.TeeReader(r, os.Stderr)
+	}
+
+	cmdPreRun(cmd)
+	if err := cmd.Start(); err != nil {
+		return "", nil, err
+	}
+
+	// We use ready to signal that postgres has started.
+	ready := make(chan bool, 1)
+	cleanup.Add(func() {
+		_ = cmd.Process.Kill()
+		_ = cmd.Wait()
+		// make sure scanner goroutine has stopped
+		<-ready
+	})
+
+	go func() {
+		defer close(ready)
+		scanner := bufio.NewScanner(r)
+		for scanner.Scan() {
+			if bytes.Contains(scanner.Bytes(), []byte("listening on Unix socket")) {
+				ready <- true
+				break
+			}
+		}
+		io.Copy(io.Discard, r) // drain r
+	}()
+
+	ok := <-ready
+	if !ok {
+		return "", nil, errors.New("failed to find ready message in log output")
+	}
+
+	// This DSN points to the directory postgres will put unix sockets.
+	return "postgresql:///postgres?host=" + pgHost, cleanup.Run, nil
+}
+
+type deferList []func()
+
+func (l *deferList) Add(f func()) {
+	*l = append(*l, f)
+}
+
+func (l *deferList) Run() {
+	for i := len(*l) - 1; i >= 0; i-- {
+		(*l)[i]()
+	}
+}
```

Comment:
I dont believe this needs to be a pointer method in this case, so this looks a tad nicer to the eye :slightly_smiling_face: 

```suggestion
func (l deferList) Run() {
	for i := len(l) - 1; i >= 0; i-- {
		l[i]()
	}
}
```

Reply:
agreed. However, I like to keep methods on a type consistent w.r.t. pointer vs value. There is a go lint tool which enforces that.